synapse/event_auth.py

```diff
@@ -734,12 +735,36 @@ def _check_power_levels(
             UserID.from_string(k)
         except Exception:
             raise SynapseError(400, "Not a valid user_id: %s" % (k,))
-
         try:
             int(v)
         except Exception:
             raise SynapseError(400, "Not a valid power level: %s" % (v,))
 
+    # Reject events with stringy power levels if required by room version
+    if (
+        event.type == EventTypes.PowerLevels
+        and room_version_obj.msc3667_int_only_power_levels
+    ):
+        for k, v in event.content.items():
+            if k in {
+                "users_default",
+                "events_default",
+                "state_default",
+                "ban",
+                "redact",
+                "kick",
+                "invite",
+            }:
+                if not isinstance(v, int):
+                    raise SynapseError(400, f"{v} must be an integer.")
+            if k in {"events", "notifications", "users"}:
+                if not isinstance(v, dict) or not all(
+                    isinstance(v, int) for v in v.values()
+                ):
+                    raise SynapseError(
+                        400, f"{v} must be a dict wherein all the values are integers."
+                    )
```

Comment:
It might be confusing to get errors like "100 must be an integer." when power levels are strings. We could use `repr` (`!r`) to format the error, so that strings will be single quoted, which is a tiny bit clearer.

```suggestion
                if not isinstance(v, int):
                    raise SynapseError(400, f"{v!r} must be an integer.")
            if k in {"events", "notifications", "users"}:
                if not isinstance(v, dict) or not all(
                    isinstance(v, int) for v in v.values()
                ):
                    raise SynapseError(
                        400, f"{v!r} must be a dict wherein all the values are integers."
                    )
```

Reply:
I haven't seen `!r` before, thanks for sharing!

Reply:
The notation was introduced in Python 3's `str.format()` and has made its way into f-strings.

https://pyformat.info/#conversion_flags describes the relevant part of `str.format`. I think most people would prefer to use f-strings these days though.